lib/internal/quic/core.js

```diff
@@ -0,0 +1,3341 @@
+    }
+
+    this.sendFD(fd, options, true);
+  };
+
+  sendFD(fd, { offset = -1, length = -1 } = {}, ownsFd = false) {
+    if (this.destroyed || this.#closed)
+      return;
+
+    validateInteger(offset, 'options.offset', { min: -1 });
+    validateInteger(length, 'options.length', { min: -1 });
+
+    if (fd instanceof fsPromisesInternal.FileHandle)
+      fd = fd.fd;
+    else if (typeof fd !== 'number')
+      throw new ERR_INVALID_ARG_TYPE('fd', ['number', 'FileHandle'], fd);
+
+    if (this.pending) {
+      return this.once('ready', () => {
+        this.sendFD(fd, { offset, length }, ownsFd);
+      });
+    }
+
+    this[kUpdateTimer]();
+    this.ownsFd = ownsFd;
+
+    // Close the writable side of the stream, but only as far as the writable
+    // stream implementation is concerned.
+    this._final = null;
+    this.end();
+
+    defaultTriggerAsyncIdScope(this[async_id_symbol],
+                               QuicStream.#startFilePipe,
+                               this, fd, offset, length);
+  }
+
+  static #startFilePipe = (stream, fd, offset, length) => {
+    const handle = new FileHandle(fd, offset, length);
+    handle.onread = QuicStream.#onPipedFileHandleRead;
+    handle.stream = stream;
+
+    const pipe = new StreamPipe(handle, stream[kHandle]);
+    pipe.onunpipe = QuicStream.#onFileUnpipe;
+    pipe.start();
+
+    // Exact length of the file doesn't matter here, since the
+    // stream is closing anyway - just use 1 to signify that
+    // a write does exist
+    stream[kTrackWriteState](stream, 1);
+  }
+
+  static #onFileUnpipe = function() {  // Called on the StreamPipe instance.
+    const stream = this.sink[owner_symbol];
+    if (stream.ownsFd)
+      this.source.close().catch(stream.destroy.bind(stream));
+    else
+      this.source.releaseFD();
+  };
+
+  static #onPipedFileHandleRead = function() {
+    const err = streamBaseState[kReadBytesOrError];
+    if (err < 0 && err !== UV_EOF) {
+      this.stream.destroy(errnoException(err, 'sendFD'));
+    }
+  };
+
+  get resetReceived() {
+    return (this.#resetCode !== undefined) ?
+      { code: this.#resetCode | 0 } :
+      undefined;
+  }
+
+  get bufferSize() {
+    // TODO(@jasnell): Implement this
+    return undefined;
+  }
+
+  get id() {
+    return this.#id;
+  }
+
+  get push_id() {
+    return this.#push_id;
+  }
+
+  close(code) {
+    this[kClose](QUIC_ERROR_APPLICATION, code);
+  }
+
+  get session() {
+    return this.#session;
+  }
+
+  _destroy(error, callback) {
+    this.#session[kRemoveStream](this);
+    const handle = this[kHandle];
+    // Do not use handle after this point as the underlying C++
+    // object has been destroyed. Any attempt to use the object
+    // will segfault and crash the process.
+    if (handle !== undefined) {
+      this.#stats = new BigInt64Array(handle.stats);
+      handle.destroy();
+    }
+    // The destroy callback must be invoked in a nextTick
+    process.nextTick(() => callback(error));
```

Comment:
I don't think `nextTick` is required?

Reply:
It is currently required but we can revisit in a subsequent PR